scertec.go

```diff
@@ -0,0 +1,189 @@
+	"errors"
+	"sync"
+	"sync/atomic"
+
+	"github.com/tailscale/setec/client/setec"
+)
+
+// Client is the setec client providing the TLS certs stored in setec by certd.
+//
+// It does not connect to the certd server directly. (in fact, certd provides no
+// cert service; it only updates TLS cert secrets in setec.)
+type Client struct {
+	sc       setec.Client
+	st       *setec.Store
+	devMode  bool        // use setec "dev/*" secrets instead of "prod/*"
+	forceRSA atomic.Bool // for testing
+
+	parsed sync.Map // secret name [string] => *parsedCert
+}
+
+// parsedCert is a cached of a previously used certificate.
+// It's used to avoid parsing the same certificate multiple times.
+// There is one parsedCert per secret name but its latest
+// pointer gets updated whenever the PEM from the secret changes.
+type parsedCert struct {
+	latest atomic.Pointer[pemAndParsed]
+}
+
+// pemAndParsed are the PEM and parsed certificate for a particular
+// version of the secret. (We don't know the secret version but
+// we notice when the PEM bytes change). The pem and parsed values
+// are immutable once set.
+type pemAndParsed struct {
+	pem    []byte
+	parsed *tls.Certificate
+	hits   atomic.Int64 // for tests
+}
+
+func secretName(devMode bool, domain string, certType string) string {
+	if devMode {
+		return "dev/certd/https-cert/" + domain + "/" + certType
+	}
+	return "prod/certd/https-cert/" + domain + "/" + certType
+}
+
+// NewClient returns a new HTTPS cert client. It blocks until all the needed
+// secrets are available for retrieval by the Secret method, or ctx ends.  The
+// context passed to NewStore is only used for initializing the store.
+//
+// As a special case for testing, if domains contains "DEV" as the last element,
+// the Client will use "dev/" secrets instead of "prod/" secrets.
+func NewClient(ctx context.Context, c setec.Client, cache setec.Cache, domains ...string) (*Client, error) {
+	var secretNames []string
+	devMode := len(domains) > 0 && domains[len(domains)-1] == "DEV"
+	if devMode {
+		domains = domains[:len(domains)-1]
+	}
+	for _, d := range domains {
+		secretNames = append(secretNames,
+			secretName(devMode, d, "rsa"),
+			secretName(devMode, d, "ecdsa"),
+		)
+	}
+	st, err := setec.NewStore(ctx, setec.StoreConfig{
+		Client:  c,
+		Secrets: secretNames,
+		Cache:   cache,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		sc:      c,
+		st:      st,
+		devMode: devMode,
+	}, nil
+}
+
+// GetCertificate returns the RSA or ECDSA certificate for hello.ServerName.
+//
+// It is the signature needed by tls.Config.GetCertificate.
+func (c *Client) GetCertificate(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	typ := "rsa"
+	if supportsECDSA(hello) && !c.forceRSA.Load() {
+		typ = "ecdsa"
+	}
+	secName := secretName(c.devMode, hello.ServerName, typ)
+	sec := c.st.Secret(secName)
+	if sec == nil {
+		return nil, errors.New("invalid server name")
+	}
+	return c.parsedCert(secName, sec.Get())
```

Comment:
The secret vends the current cached copy from the Store, so it only exchanges data with the server when it explicitly polls for an update. The Store already includes the conditional check so that the server will only vend new bytes if the version has changed from the one the store already has cached.